### PR TITLE
[LPT][WIP] Support of INT8 inference of LSTM layer

### PR DESCRIPTION
**Pull request by ndemashov**
### Details:
 - Support of INT8 inference of LSTM layer

### Tickets:
 - 56479

### CODE REVIEW

File: src/common/low_precision_transformations/src/lstm.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    const auto fq_W = ngraph::pattern::wrap_type<opset1::FakeQuantize>({ W,
+        input_low,
+        input_high,
+        output_low,
+        output_high });
+    const auto fq_R = ngraph::pattern::wrap_type<opset1::FakeQuantize>({ R,
+        input_low,
+        input_high,
+        output_low,
+        output_high });
```
Comment (eshoguli) [inline src/common/low_precision_transformations/src/lstm.cpp:55]: 1) Don't use the same `Constant` operation instances for different `FakeQuantize`. You build function with four `FakeQuantize` with the same four `Constant`.
2) Create and reuse helper: `wrap_fake_quantize`

**Review by eshoguli — COMMENTED** [review]
There are comments, tests review is still in progress.

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2018-2022 Intel Corporation
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp:1]: done, thank you

File: src/common/low_precision_transformations/include/low_precision/lstm.hpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <memory>
+#include <ngraph/ngraph.hpp>
+#include "low_precision/layer_transformation.hpp"
+
+namespace ngraph {
+namespace pass {
+namespace low_precision {
+
+class LP_TRANSFORMATIONS_API LSTMTransformation : public LayerTransformation {
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/lstm.hpp]: done, thank you

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+class LP_TRANSFORMATIONS_API SkipCleanupAttribute : public SharedAttribute<bool> {
+public:
+    OPENVINO_RTTI("LowPrecision::SkipCleanup", "", ov::RuntimeAttribute, 0);
+    SkipCleanupAttribute(const bool skip);
+
+    static ov::Any create(const std::shared_ptr<ngraph::Node>& node, const bool skip);
+    // vizualize shared attributes details in VizualizeTree pass
+    std::string to_string() const override;
+};
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp]: done, thank you

File: src/common/low_precision_transformations/src/lstm.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/opsets/opset1.hpp>
+
+#include <memory>
+#include <ngraph/node.hpp>
+#include <ngraph/opsets/opset1.hpp>
+#include <ngraph/opsets/opset5.hpp>
+#include <ngraph/pattern/op/or.hpp>
+
+#include "low_precision/concat.hpp"
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/lstm.cpp]: done, thank you

File: src/common/low_precision_transformations/src/lstm.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+bool LSTMTransformation::transform(TransformationContext& context, ngraph::pattern::Matcher& m) {
+     const auto lstm = m.get_match_root();
+    if (!canBeTransformed(context, lstm)) {
+        return false;
+    }
+    OutputVector inputs;
+    for (size_t parentIndex = 0ul; parentIndex < lstm->get_input_size(); parentIndex++) {
+        inputs.push_back(lstm->get_input_node_shared_ptr(parentIndex));
+    }
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/lstm.cpp]: done, thank you

File: src/common/low_precision_transformations/src/lstm.cpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    const auto H = ngraph::pattern::wrap_type<ngraph::opset5::Parameter>();
+    const auto C = ngraph::pattern::wrap_type<ngraph::opset5::Parameter>();
+    const auto W = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+    const auto R = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+    const auto B = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+
+    const auto input_low = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+    const auto input_high = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+    const auto output_low = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
+    const auto output_high = ngraph::pattern::wrap_type<ngraph::opset5::Constant>();
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/lstm.cpp]: done, thank you

File: src/common/low_precision_transformations/src/lstm.cpp [hunk 8]
```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (C) 2018-2022 Intel Corporation
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/lstm.cpp]: done, thank you

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "low_precision/lpt_visibility.hpp"
+#include "low_precision/rt_info/attribute_parameters.hpp"
+#include "low_precision/rt_info/shared_value_attribute.hpp"
+
+namespace ngraph {
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief PrecisionsAttribute defines precision which is required for input/output port or an operation.
+ */
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp]: done, thank you

File: src/common/low_precision_transformations/src/lstm.cpp (hunk 1 above)
Comment (ndemashov) [reply src/common/low_precision_transformations/src/lstm.cpp]: done, thank you

File: src/common/low_precision_transformations/src/fuse_multiply_to_fake_quantize.cpp [hunk 10]
```diff
@@ -112,6 +114,13 @@ bool FuseMultiplyToFakeQuantizeTransformation::canBeTransformed(const Transforma
         return false;
     }
 
+    auto skip = getAttribute<SkipCleanupAttribute>(operation);
+    if (!skip.empty()) {
+        if (skip.as<SkipCleanupAttribute>().value()) {
+            return false;
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/fuse_multiply_to_fake_quantize.cpp]: done, thank you

File: src/common/low_precision_transformations/src/fuse_subtract_to_fake_quantize.cpp [hunk 11]
```diff
@@ -116,6 +117,12 @@ bool FuseSubtractToFakeQuantizeTransformation::canBeTransformed(const Transforma
     if (fq->get_output_target_inputs(0).size() != 1) {
         return false;
     }
+    auto skip = getAttribute<SkipCleanupAttribute>(operation);
+    if (!skip.empty()) {
+        if (skip.as<SkipCleanupAttribute>().value()) {
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/fuse_subtract_to_fake_quantize.cpp]: done, thank you

File: src/common/low_precision_transformations/src/recurrent_cell.cpp [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        {dequantization_convert_X, multiply_constant});
+    const auto dequantization_multiply_H = ngraph::pattern::wrap_type<ngraph::opset1::Multiply>(
+        {dequantization_subtract_H, multiply_constant});
+    const auto dequantization_multiply_without_subtract_H = ngraph::pattern::wrap_type<ngraph::opset1::Multiply>(
+        {dequantization_convert_H, multiply_constant});
+
+    const auto lstm_cell = ngraph::pattern::wrap_type<ngraph::opset5::LSTMCell>(
+        {fq_X, fq_H, C, fq_W, fq_R, B});
+    const auto lstm_cell_with_dequantizations = ngraph::pattern::wrap_type<ngraph::opset5::LSTMCell>(
+        {dequantization_multiply_X, dequantization_multiply_H, C, fq_W, fq_R, B});
```
Comment (eshoguli) [inline src/common/low_precision_transformations/src/recurrent_cell.cpp:57]: Refactoring idea: not clear source code with copy/paste. Can we refactor wrapper creation for dequantization patter - extract to new function `wrap_dequantization`, please? Like as:
``` cpp
const auto lstm_cell_with_dequantizations = ngraph::pattern::wrap_type<ngraph::opset5::LSTMCell>({wrap_dequantization(X), wrap_dequantization(H), C, wrap_fake_quantize(W), wrap_fake_quantize(R), B});
```

**Review by eshoguli — COMMENTED** [review]
minor comments, please fix

File: src/common/low_precision_transformations/src/recurrent_cell.cpp [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    const auto lstm_cell_with_dequantizations_without_subtract = ngraph::pattern::wrap_type<ngraph::opset5::LSTMCell>(
+        {dequantization_multiply_without_subtract_X, dequantization_multiply_without_subtract_H, C, fq_W, fq_R, B});
+
+    const auto seq_lengths = ngraph::pattern::wrap_type<ngraph::opset1::Constant>();
+    const auto lstm_sequence = ngraph::pattern::wrap_type<ngraph::opset5::LSTMSequence>(
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/recurrent_cell.cpp]: done, thank you

File: src/tests/functional/inference_engine/lp_transformations/recurrent_cell_transformation.cpp [hunk 14]
```diff
@@ -0,0 +1,608 @@
...
+        if (!testValues.result.dequantizationAfter.multiply.empty()) {
+            testValues.result.dequantizationAfter.multiply.outPrecision = precision;
+        }
+
+        if (!testValues.params.updatePrecisions && (precision == ngraph::element::f32) &&
+            !testValues.result.dequantizationAfter.convert.empty()) {
+            testValues.result.dequantizationAfter.convert = {};
+        }
+
+        IntervalsAlignmentSharedValue::Interval interval{-1.28f, 2.55f};
```
Comment (eshoguli) [inline src/tests/functional/inference_engine/lp_transformations/recurrent_cell_transformation.cpp:151]: not used, please remove

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <set>
+#include <unordered_set>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp]: done, thank you

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <ngraph/node.hpp>
+#include <ngraph/pass/graph_rewrite.hpp>
+#include <ngraph/variant.hpp>
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp]: done, thank you

File: src/plugins/intel_cpu/src/nodes/tensoriterator.h [hunk 17]
```diff
@@ -108,6 +108,13 @@ class TensorIterator : public Node {
 
     void setExtManager(const ExtensionManager::Ptr& extMgr) { ext_mng = extMgr; }
 
+    Graph getSubGraph() const {
+        return sub_graph;
+    }
+    std::shared_ptr<ov::Node> getOriginalOp() const {
+        return ngraphOp;
+    }
+
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/plugins/intel_cpu/src/nodes/tensoriterator.h]: done, thank you

File: src/tests/functional/inference_engine/lp_transformations/recurrent_cell_transformation.cpp (hunk 14 above)
Comment (ndemashov) [reply src/tests/functional/inference_engine/lp_transformations/recurrent_cell_transformation.cpp]: done, thank you

File: src/tests/functional/plugin/cpu/shared_tests_instances/low_precision_transformations/recurrent_cell_transformation.cpp [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+        },
+        // W
+        {255ul, {}, {0.f}, {2.55f}, {0.f}, {2.55f}},
+        {},
+        {{}, {}, {}},
+        // R
+        {255ul, {}, {0.f}, {2.55f}, {0.f}, {2.55f}},
+        {},
+        {{}, {}, {}},
+        ngraph::builder::subgraph::RecurrentCellFunction::RNNType::LSTMSequence,
```
Comment (eshoguli) [inline src/tests/functional/plugin/cpu/shared_tests_instances/low_precision_transformations/recurrent_cell_transformation.cpp:129]: use `LSTMCell`, please

Comment (ndemashov) [reply src/tests/functional/plugin/cpu/shared_tests_instances/low_precision_transformations/recurrent_cell_transformation.cpp]: done, thank you

File: src/common/low_precision_transformations/src/recurrent_cell.cpp [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+                                                                  updatePrecisions);
+                std::shared_ptr<ngraph::Node> new_fq = std::get<0>(QDQ);
+                std::shared_ptr<ngraph::Node> deq_multiply = std::get<1>(QDQ);
+                if (deq_multiply == nullptr || new_fq == nullptr) {
+                    return false;
+                }
+                auto multiply_parent = deq_multiply->get_input_node_shared_ptr(0);
+                if (is_type<ngraph::opset1::Subtract>(multiply_parent)) {
+                    return false;
+                }
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/recurrent_cell.cpp]: done, thank you

[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/src/recurrent_cell.cpp]: done, thank you

File: src/common/low_precision_transformations/src/recurrent_cell.cpp (hunk 12 above)
Comment (ndemashov) [reply src/common/low_precision_transformations/src/recurrent_cell.cpp]: done, thank you. Also added wrap_quantization.

File: src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "low_precision/lpt_visibility.hpp"
+#include "low_precision/rt_info/attribute_parameters.hpp"
```
[the remark this answers is not in the document]
Comment (ndemashov) [reply src/common/low_precision_transformations/include/low_precision/rt_info/skip_cleanup_attribute.hpp:9]: done, thank you. #include "low_precision/rt_info/attribute_parameters.hpp" is necessary

**Comment by dmitry-gorokhov** [comment]
closed as outdated

**Pull request closed without merging**